sled-agent/src/sled_agent.rs

```diff
@@ -326,12 +326,22 @@ impl SledAgent {
     /// Ensures that particular services should be initialized.
     ///
     /// These services will be instantiated by this function, will be recorded
-    /// to a local file to ensure they start automatically on next boot.
+    /// to a local file to ensure they start automatically on next boot. After
+    /// being recorded to the filesystem, Nexus will be notified
     pub async fn services_ensure(
         &self,
         requested_services: ServiceEnsureBody,
     ) -> Result<(), Error> {
-        self.services.ensure(requested_services).await?;
+        self.services.ensure(requested_services.clone()).await?;
+
+        self.services
+            .notify_running_services(
+                self.id,
+                self.lazy_nexus_client.clone(),
+                requested_services,
+            )
+            .await?;
```

Comment:
As discussed offline, I do think there is a weird inversion of control here, but I'm not sure it should block this PR.

I expect that generally, Nexus will be the entity responsible for deciding "what services should go where", and should be making calls to `services_ensure` here. If Nexus updates the DB, calls the sled agent (which causes `services_ensure` to be invoked), and the sled agent calls back up into Nexus, which could update the DB, we end up having "multiple sources of truth" for what should be in the DB. I could see this getting pretty confusing, and it basically boils down to:

- Should the DB representation of services describe what Nexus *wants* to provision on sleds?
or
- Should the DB representation of services describe what *currently* exists on sleds?

By attempting to describe both, I think we end up in a sort of odd split-brain situation

Reply:
In this future, it should only be `the DB representation of services describe what Nexus wants to provision on sleds` - said another way, it should only be what should exist, not what does.

This is similar to DNS: as 248 says, "Service discovery is separate from connection management.". I imagine DNS records could be populated from this service table too if the internal DNS service bounces and loses the in-memory DB, though that's a separate issue.

Reply:
Honestly I'm still hung up on this - I think it's the wrong flow of control for the long-term, and I don't understand what short-term problem we're fixing with this PR.

You said something to me in a chat that:

> I need the pantry to register itself when it boots so I can later look up the address

But you should be able to create a service, add a DNS record, and look up the DNS name *without* "instructing" Nexus what services should exist.

The rack setup service does the following:

It infers the addresses of DNS servers based on the AZ subnet:
https://github.com/oxidecomputer/omicron/blob/e664011b3934ba9ec495774e9439d76baf516d5e/sled-agent/src/rack_setup/service.rs#L603-L606

When initializing services, it inserts DNS records for them:
https://github.com/oxidecomputer/omicron/blob/e664011b3934ba9ec495774e9439d76baf516d5e/sled-agent/src/rack_setup/service.rs#L623-L634

I'm not clear on how, where, or when you're launching the pantry, but service discovery via DNS is the mechanism that we should be using (and it's the mechanism we *are* using to find services, for example, it's how [sleds find nexus](https://github.com/oxidecomputer/omicron/blob/main/sled-agent/src/nexus.rs)).

Reply:
> I'm not clear on how, where, or when you're launching the pantry...

FWIW, the Pantry will be launched as a separate service zone.

> But you should be able to create a service, add a DNS record, and look up the DNS name without "instructing" Nexus what services should exist.

I agree that DNS should be the method of service discovery. With this PR I intending it to be "tell Nexus what services *are* running", where it seems like the `services` table is (and should probably only be) "what services *should* be running". My mistake.

Based on this, I also agree that wiring up `PUT /service` is the wrong thing to do. Should I close this PR, or change it to rip out the endpoint?